core/commands/key.go

```diff
@@ -0,0 +1,328 @@
+package commands
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	"github.com/ipfs/go-ipfs/core/corenet"
+	crypt "github.com/ipfs/go-ipfs/crypt"
+	ci "github.com/ipfs/go-ipfs/p2p/crypto"
+	peer "github.com/ipfs/go-ipfs/p2p/peer"
+)
+
+const PeerKeyName = "local"
+
+var KeyCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Interact with IPFS keypairs",
+		ShortDescription: `
+'ipfs key' is a command used to manage ipfs keypairs.
+`,
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"gen":  keyGenCmd,
+		"send": keySendCmd,
+		"recv": keyRecvCmd,
+	},
+}
+
+type KeyOutput struct {
+	Keyname string
+	KeyID   string
+	Bits    int
+	Type    string
+}
+
+var keyGenCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Generate a new named ipfs keypair",
+		ShortDescription: `
+'ipfs key gen' is a command used to generate new keypairs.
+If any options are not given, the command will go into interactive mode and prompt
+the user for the missing fields.
+`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("keyname", true, false, "The name of the key to create"),
+	},
+	Options: []cmds.Option{
+		cmds.IntOption("b", "bits", "bitsize of key to generate (default = 2048)"),
+		cmds.StringOption("t", "type", "type of key to generate (default = RSA)"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		nd, err := req.Context().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		kname := req.Arguments()[0]
+		if kname == PeerKeyName {
+			res.SetError(errors.New("cannot name key 'local', would overwrite peer key"), cmds.ErrNormal)
+			return
+		}
+
+		_, err = nd.Repo.Keystore().GetKey(kname)
+		if err != nil && !os.IsNotExist(err) {
+			res.SetError(fmt.Errorf("error checking for key in keystore: %s", err), cmds.ErrNormal)
+			return
+		}
+		if err == nil {
+			res.SetError(errors.New("key with that name already exists"), cmds.ErrNormal)
+			return
+		}
+
+		bits, found, err := req.Option("b").Int()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		if !found {
+			bits = 2048
+		}
+
+		typ, found, err := req.Option("t").String()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		if !found {
+			typ = "RSA"
+		}
+
+		var ntyp int
+		switch typ {
+		case "RSA":
+			ntyp = ci.RSA
```

Comment:
we had a PR a while back adding in ed25519 support. It didnt get merged for one reason or another. I'm not sold on RSA, but its easy to use for now, and it works. I'm making things modular so we can change it later easily.